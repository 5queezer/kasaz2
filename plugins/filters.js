import Vue from 'vue'

Vue.filter('toLocale', function (value, unit) {
  if (!value) { return '' }
  value = value.toLocaleString()
  if (unit) {
    value += ' ' + unit
  }
  return value
})
