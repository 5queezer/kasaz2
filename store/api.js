import { setup } from 'axios-cache-adapter'

const api = setup({
  // baseURL: 'http://localhost:43287/',
  cache: {
    maxAge: 15 * 60 * 1000
  }
})

function getApartments (params) {
  return api.get('/response.json', { params })
    .then((response) => {
      return response.data.markers
    })
}

export default {
  getApartments
}
