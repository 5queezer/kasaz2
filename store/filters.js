import deepCleaner from 'deep-cleaner'
import { mergeDeep } from '@/assets/utils'

export const state = () => ({
  // init state with reset first
  // filters: {},
  // locale: '',
  // location: {},
  // viewport: {}
})
export const getters = {
  get (state) {
    return state
  }
}
export const mutations = {
  set (state, param) {
    if (param.key && typeof param.value !== 'undefined') {
      // key as string of object structure
      // e.g. filters[price][max]
      const value = param.value
      const command = param.key.split(']').join('').split(/\[+/)
      const last = command.pop()
      let key = state
      command.forEach((attribute) => {
        if (!key[attribute]) {
          key.attribute = {}
        }
        key = key[attribute]
      })
      if (value === null) {
        delete key[last]
      } else {
        key[last] = value
      }
    } else {
      for (const [key, value] of Object.entries(param)) {
        if (!state[key]) {
          state.key = value
        } else {
          state[key] = mergeDeep(state[key], value)
        }
      }
      deepCleaner(state)
    }
  },
  reset (state) {
    // filter scheme
    const init = {
      filters: {
        lodgingType: {
          attic: true,
          duplex: true,
          flat: true,
          ground_floor: true,
          house: true,
          loft: true,
          studio: true
        },
        showHiddenListings: false,
        sortby: 'relevance',
        price: {
          min: undefined,
          max: undefined
        },
        surface: {
          min: undefined,
          max: undefined
        },
        bedrooms: undefined,
        bathrooms: undefined
      },
      locale: 'es',
      location: { city: 'Barcelona' },
      viewport: {
        neLat: undefined,
        neLng: undefined,
        swLat: undefined,
        swLng: undefined
      }
    }
    Object.freeze(init)

    for (const [key, value] of Object.entries(init)) {
      state[key] = value
    }
  }
}
