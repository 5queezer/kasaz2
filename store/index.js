import api from '@/assets/api'

export const state = () => ({
  data: [],
  loading: false,
  index: undefined,
  perPage: 12
})

export const getters = {
  data (state) {
    return state.data
  },
  loading (state) {
    return state.loading
  },
  currentPage (state) {
    return state.index ? parseInt(state.index / state.perPage) + 1 : 1
  },
  perPage (state) {
    return state.perPage
  },
  paginated (state, getters) {
    const start = (getters.currentPage - 1) * state.perPage
    const end = start + state.perPage
    const max = state.data.length - 1
    if (max === start) { return [start] }
    return getters.data.slice(start, end > max ? max + 1 : end)
  },
  count (state) {
    return state.data.length
  },
  getIndex (state) {
    return state.index || 0
  },
  current (state, getters) {
    return getters.data[getters.getIndex] || {}
  }
}

export const mutations = {
  add (state, apartment) {
    if (typeof apartment.l === 'undefined' || typeof apartment.g === 'undefined') {
      throw new TypeError(`malformed geo location { lat: ${apartment.l}, lng: ${apartment.g} }`)
    }
    if (typeof apartment.id === 'undefined') {
      apartment.id = Math.max(state.data.map(s => s.id)) + 1
    } else {
      for (const entry of state.data) {
        if (entry.id === apartment.id) { throw new Error('double entry') }
      }
    }

    state.data.push(apartment)
  },
  reset (state) {
    state.data = []
    state.index = undefined
  },
  save (state, data) {
    state.data = data
  },
  loading (state, value) {
    state.loading = !!value
  },
  activate (state, id) {
    state.index = state.data.findIndex(item => item.id === id)
  },
  setPage (state, page) {
    const startIndex = state.perPage * (page - 1)
    const offset = (state.index || 0) % 10
    const max = state.data.length - 1
    const endIndex = startIndex + offset
    state.index = endIndex > max ? max : endIndex
  },
  next (state) {
    const index = state.index + state.perPage
    const max = state.data.length - 1
    state.index = index > max ? max : index
  },
  prev (state) {
    const index = state.index - state.perPage
    state.index = index < 0 ? 0 : index
  },
  begin (state) {
    state.index = state.data.length > 0 ? 0 : undefined
  },
  end (state) {
    const max = state.data.length - 1
    state.index = max
  }

}

export const actions = {
  async fetch ({ commit }, params) {
    commit('loading', true)
    const result = await api.getApartments(params)
    commit('save', result)
    commit('loading', false)
    return result.length > 0
  }
}
