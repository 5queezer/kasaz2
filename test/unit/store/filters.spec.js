import { mutations, state } from '@/store/filters'

describe('Filter Mutations', () => {
  beforeEach(() => {
    mutations.reset(state)
  })

  it('can set a filter with key and value', () => {
    const key = 'filters[price][max]'
    const value = 1e6
    mutations.set(state, { key, value })
    expect(state.filters.price).toHaveProperty('max', value)
  })

  it('can set a filter with an object', () => {
    const newValue = 1e6
    const filters = {
      price: {
        max: newValue
      }
    }
    mutations.set(state, { filters })
    expect(state.filters.price).toHaveProperty('max', newValue)
  })

  it('can remove a filter', () => {
    const key = 'filters[price][max]'
    const value = 1e6
    mutations.set(state, { key, value })
    mutations.set(state, { key, value: undefined })
    expect(state).not.toHaveProperty(name)
  })

  it('can remove a filter object-style', () => {
    const filters = {
      price: {
        max: 1e6
      }
    }
    mutations.set(state, { filters })
    filters.price.max = undefined
    mutations.set(state, { filters })
    expect(state.filters.price).not.toHaveProperty('max')
  })
})
