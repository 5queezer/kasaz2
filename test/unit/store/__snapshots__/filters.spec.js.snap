// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Mutations can reset the filter to standard values 1`] = `
Object {
  "bathrooms": undefined,
  "bedrooms": undefined,
  "lodgingType": Object {
    "attic": true,
    "duplex": true,
    "flat": true,
    "ground_floor": true,
    "house": true,
    "loft": true,
    "studio": true,
  },
  "price": Object {
    "max": undefined,
    "min": undefined,
  },
  "showHiddenListings": false,
  "sortby": "relevance",
  "surface": Object {
    "max": undefined,
    "min": undefined,
  },
}
`;

exports[`Filter Mutations can reset the filter to standard values 2`] = `
Object {
  "neLat": undefined,
  "neLng": undefined,
  "swLat": undefined,
  "swLng": undefined,
}
`;

exports[`Filter Mutations can reset the filter to standard values 3`] = `
Object {
  "city": "Barcelona",
}
`;

exports[`Filter Mutations can reset the filter to standard values 4`] = `"es"`;
